### PR TITLE
PEP 581, 588: Split GitHub migration rationale and plan into two separate PEPs

### PR DESCRIPTION
**Pull request by warsaw**

### CODE REVIEW

File: pep-0581.rst [hunk 1]
```diff
@@ -69,29 +74,33 @@ not currently available on Roundup / bpo.
...
-  existing maintainers in terms of reviewing, testing, and applying patches.
+- The upstream Roundup code is in Mercurial. Without any CI available,
+  it puts heavy burden on the few existing maintainers in terms of
+  reviewing, testing, and applying patches.
 
-- At its current state, it is not equipped to accept lots of contributions from
-  people who aren't already familiar with the code base.
+  There is an open discussion about moving the source code of bpo to
+  GitHub [#]_. If the source code of bpo does move to GitHub, it will
+  become difficult to update patches from upstream. But as long as it
```
Comment (berkerpeksag) [inline pep-0581.rst]: This isn't a huge problem. Roundup has recently added builtin REST interface. We can start to port our GitHub integration to use it. I don't think we are relaying on our fork's customizations too much. As always, I'm volunteering to do this work.

File: pep-0581.rst [hunk 2]
```diff
@@ -69,29 +74,33 @@ not currently available on Roundup / bpo.
...
+
 
 Issues with Roundup / bpo
 -------------------------
 
 - Less than five people maintain bpo. Some of them are core developers.
 
-- It is in Mercurial. Without any CI available, it puts heavy burden on the few
-  existing maintainers in terms of reviewing, testing, and applying patches.
+- The upstream Roundup code is in Mercurial. Without any CI available,
```
Comment (berkerpeksag) [inline pep-0581.rst]: Personally, I don't see how this is a problem here. `pip install` supports installing from Mercurial repositories and its testing suite isn't that hard to run. Plus, as I said below, our Roundup instance can easily be moved to GitHub.

File: pep-0581.rst [hunk 3]
```diff
@@ -69,29 +74,33 @@ not currently available on Roundup / bpo.
...
+  is in Mercurial, it is difficult to maintain and onboard new
+  contributors.
 
-- The upstream Roundup is in Mercurial. There is an open discussion about
-  moving the source code of bpo to GitHub [#]_. If the source code of
-  bpo does move to GitHub, it will become difficult to update patches from
-  upstream. But as long as it is in Mercurial, it is difficult to maintain
-  and onboard new contributors.
+- In its current state, the project is not equipped to accept lots of
+  contributions from people who aren't already familiar with the code
```
Comment (berkerpeksag) [inline pep-0581.rst]: I find the most complicated part of the installation process is installation of our Roundup fork. Once we get rid of our fork, it won't be harder than installing python.org.

I'd share my `INSTALL` document here, but unfortunately I left it in my computer in my hometown. Buy you can take a look at the Docker image at https://github.com/python/docker-bpo.

File: pep-0581.rst [hunk 4]
```diff
@@ -69,29 +74,33 @@ not currently available on Roundup / bpo.
...
+  base.
 
 - The user interface needs an update and redesign. It will require UX/UI research
   to keep it up to date with current web standards, including accessibility.
 
-- Email address is exposed. There is no choice to mask it.
+- Users email addresses are exposed. There is no option to mask it.
```
Comment (berkerpeksag) [inline pep-0581.rst]: Roundup can be modified to mask usernames and/or email addresses. See https://issues.roundup-tracker.org/issue2551034 for example: Usernames masked as `[hidden]` there.

File: pep-0581.rst [hunk 5]
```diff
@@ -105,25 +114,29 @@ Issues with Roundup / bpo
...
 Using another issue tracker will require yet another learning curve, for having
 to learn and get used to a different interface. We'll also need to learn and
 figure out how to build the integrations with GitHub.
 
-By using GitHub issues, where the CPython source code is currently hosted and where
-pull requests are taking place, we'll be providing consistent experience to
-contributors and maintainers, while not having to jump from one interface to another.
+By using GitHub issues, where the CPython source code is currently
+hosted and where pull requests are taking place, we'll be providing
+consistent experience to contributors and maintainers, while not
```
Comment (berkerpeksag) [inline pep-0581.rst]: We are not the only project that uses a different issue tracker while using GitHub to host their source code. Some projects (SQLAlchemy) even use GitLab for issue tracking.

**Comment by berkerpeksag** [comment]
Barry told me to post my [email](https://mail.python.org/pipermail/python-committers/2019-April/006696.html) in python-committers here. I'll try to summarize it.

I'm strongly against using GitHub Issues because of the following reasons:

* Github doesn't listen to its users feedback. When you reach out to them, you usually get a "we can't say whether we are going to do this or not" response. Instead, they are busy with adding [more important features](https://help.github.com/en/articles/customizing-your-repositorys-social-media-preview).
* See issues [Allow for the tweaking of the commit title template (or at least to use `GH` over `#`](https://github.com/maintainers/early-access-feedback/issues/77), [Don't populate (or make it optional) intermediate commits automatically in commit message](https://github.com/maintainers/early-access-feedback/issues/153) and [can the preserve the author of the original commit in the squash merge?](https://github.com/python/miss-islington/issues/16) for more examples. Note that the last one is actually a bug.
* We are trying to workaround GitHub's lack of features by creating custom tools. https://github.com/berkerpeksag/cpython-emailer-webhook is a great example to this. I'd rather spend my free time to maintain our infrastructure than trying to workaround a company's product.
* Maintaining our tools that communicate with GitHub also take time. See https://github.com/python/bedevere/pull/163 for a recent example.

**Comment by brettcannon** [comment]
@berkerpeksag can I request you tweak the tone of your comment? It's coming off as unnecessarily hostile.

**Comment by berkerpeksag** [comment]
@brettcannon which part of my comment do you think is hostile?

**Comment by brettcannon** [comment]
@berkerpeksag primarily your first bullet point. While _you_ might think GitHub doesn't listen, _I_ don't think that's true but you state it like it's a universal truth. And your last sentence in that point very much comes off as a dig as well. And just because they don't announce features before they launch doesn't mean they don't listen (I say the same things to my users at my work because I know if I over-promise and under-deliver I'm the one they will yell at, so it's easier to just not say anything until something is already released).

**Comment by berkerpeksag** [comment]
@brettcannon I gave two examples from my own experience that they didn't listen (both issues can be classified as bug) in the second bullet point. I'd be happy to combine the first two bullet points if you think that would make my comment better.

> And just because they don't announce features before they launch doesn't mean they don't listen

True, but none of my reports were actually feature requests and I definitely don't expect constant communication from them. I don't think it's unreasonable to expect a response like "we are going to fix this in X months" a year after I reported it in a place where they were created to gather feedback from the community.

**Comment by Mariatta** [comment]
> Github doesn't listen to its users feedback. When you reach out to them, you usually get a "we can't say whether we are going to do this or not" response. Instead, they are busy with adding more important features.

Sorry you had terrible experience, but I think you're generalizing the situation. Personally I've had good experiences with GitHub. Whenever I had issues and questions, they answered within 24 hours. So I think saying "GitHub doesn't listen" doesn't sound fair, or maybe I'm really lucky.

> See issues [Allow for the tweaking of the commit title template (or at least to use GH over #](https://github.com/maintainers/early-access-feedback/issues/77), [Don't populate (or make it optional) intermediate commits automatically in commit message](https://github.com/maintainers/early-access-feedback/issues/153) and [can the preserve the author of the original commit in the squash merge?](https://github.com/python/miss-islington/issues/16) for more examples. Note that the last one is actually a bug.

I feel these issues are not related to GitHub-as-issue-tracker, and not the problems that PEP 581 trying to solve, so it seems like a distraction to current topic. I also don't feel the last one as a bug. If miss-islington actually did the work to backport and create the PR, it can earn the credit IMO. The others problems you mentioned here, we've tried to help solve by miss-islington's automerge. If core devs still chose not to use the automerge feature, then I don't know what to say to that...  

> I'd rather spend my free time to maintain our infrastructure than trying to workaround a company's product.

You're definitely free to choose how you spend your free time. Roundup itself is not going away. It's not officially maintained by Python core developers, although several core developers like yourself maintain it. (thank you!!). I can understand if you're not interested in maintaining other workarounds like bedevere, miss-islington. However these projects are on GitHub and open source, and we already have quite a number of contributions to them from non core developers, so I don't feel there will be an issue. 

> We are trying to workaround GitHub's lack of features by creating custom tools. 

Personally, I think that's the purpose of the APIs, and I actually think it is great that GitHub has a set of nice features, and we can still further customize the experience by building our own tools by utilizing their APIs.

File: pep-0588.rst [hunk 6]
```diff
@@ -10,7 +11,269 @@ Created: 27-Mar-2019
...
+--------------------------------
+
+The bug triagers on bpo are valuable to the core Python workflow, and we
+definitely would need even more help with triaging issues on GitHub.
+
+It has been `proposed on Discourse
+<https://discuss.python.org/t/proposal-create-bug-triage-team-on-github/992/5>`_
+for us to create a "bug triage" team on GitHub to help with closing
+issues, notifying the appropriate parties, as well as applying labels
+to issues and pull requests. We can grant the "write" permission to
```
Comment (ammaraskar) [inline pep-0588.rst]: This should be updated to reflect that Github now has a triage level permission https://discuss.python.org/t/pep-581-using-github-issues/535/36?u=ammaraskar

Comment (pradyunsg) [reply pep-0588.rst]: It's a private beta though.

**Pull request merged**